### PR TITLE
Add back channel authentication

### PR DESCRIPTION
**Pull request by Dilhasha**
This PR adds back channel authentication for Identity cloud's Admin application.

### CODE REVIEW

File: jaggeryapps/admin/acs.jag [hunk 1]
```diff
@@ -114,17 +115,31 @@ if (samlResponse != null) {
...
+            backChannelAuthHandler.invalidateSession(issuer, samlRequest);
+            invalidated = true;
+        }
+    }
+    /*------------------------ handle back channel logout request end --------------------*/
+    if (!invalidated) {
+        var index = sso.client.decodeSAMLLogoutRequest(sso.client.getSamlObject(samlRequest));
 
-    log.info("Session index of the SLO request is " + index);
+        log.info("Session index of the SLO request is " + index);
```
Comment (danushkaf) [inline jaggeryapps/admin/acs.jag]: Make this a debug log.

File: jaggeryapps/admin/acs.jag [hunk 2]
```diff
@@ -114,17 +115,31 @@ if (samlResponse != null) {
...
 
-    var httpSessionForSessionIndex = application.get('httpSessions')[index];
+        var httpSessionForSessionIndex = application.get('httpSessions')[index];
 
-    if(httpSessionForSessionIndex){
-        log.info("Invalidating http session for the session index " + index);
-        httpSessionForSessionIndex.invalidate();
-        delete application.get('httpSessions')[index];
+        if (httpSessionForSessionIndex) {
+            log.info("Invalidating http session for the session index " + index);
```
Comment (danushkaf) [inline jaggeryapps/admin/acs.jag]: Make this a debug log.

File: jaggeryapps/admin/authentication/auth_config.json [hunk 3]
```diff
@@ -14,5 +14,15 @@
 		"SAML.PrivateKeyPassword" : "wso2carbon",
 		"SAML.KeyStore" : "{carbon.home}/repository/resources/security/wso2carbon.jks"
 		
+	},
+	"backChannelAuth" : {
+		"enabled" : true,
+		"issuer" : "wso2.my.dashboard",
+		"tokenEP" : "https://identity.cloudlocal.wso2.com:9443/oauth2/token",
```
Comment (danushkaf) [inline jaggeryapps/admin/authentication/auth_config.json]: better to make all urls localhost in committed file.

File: jaggeryapps/admin/login.jag [hunk 4]
```diff
@@ -4,17 +4,23 @@ var util = require("util/utility.jag");
...
 function login(){
     util.setServerURL("true");
 	initAuthenticationProtocol();
 	
 	var user = session.get(LOGGED_IN_USER);	
     if(user == null){
-    	
+    	//Check whether back chanel auth can be done
+    	var success = checkBackChannelAuth();
+    	if (success) {
```
Comment (danushkaf) [inline jaggeryapps/admin/login.jag]: What about 
if (checkBackChannelAuth()) {

File: jaggeryapps/admin/login.jag [hunk 5]
```diff
@@ -39,8 +45,135 @@ lg.info('saml not enabled');
     }
 }
 
+function canAuthenticate(authCookie, samlssoTokenId) {
+	var checkSSO = authConfig.EnableSAMLSSOLogin;
+	if (!checkSSO || session.get("LOGGED_IN_USER")) {
```
Comment (danushkaf) [inline jaggeryapps/admin/login.jag]: Again do we need checkSSO variable?

**Review by danushkaf — CHANGES_REQUESTED** [review]

File: jaggeryapps/admin/authentication/auth_config.json [hunk 6]
```diff
@@ -14,5 +14,15 @@
 		"SAML.PrivateKeyPassword" : "wso2carbon",
 		"SAML.KeyStore" : "{carbon.home}/repository/resources/security/wso2carbon.jks"
 		
+	},
+	"backChannelAuth" : {
+		"enabled" : true,
```
Comment (malithie) [inline jaggeryapps/admin/authentication/auth_config.json]: Make default enabled false as we cannot directly use this in local distribution

**Review by malithie — CHANGES_REQUESTED** [review]

File: jaggeryapps/admin/acs.jag [hunk 7]
```diff
@@ -114,17 +115,34 @@ if (samlResponse != null) {
...
-
-    log.info("Session index of the SLO request is " + index);
-
-    var httpSessionForSessionIndex = application.get('httpSessions')[index];
-
-    if(httpSessionForSessionIndex){
-        log.info("Invalidating http session for the session index " + index);
-        httpSessionForSessionIndex.invalidate();
-        delete application.get('httpSessions')[index];
+    log.info("SLO request received.");
```
Comment (danushkaf) [inline jaggeryapps/admin/acs.jag]: Reason for this being in info level? better to add debug data to log and make it debug

File: jaggeryapps/admin/acs.jag [hunk 8]
```diff
@@ -114,17 +115,34 @@ if (samlResponse != null) {
...
+    var invalidated = false;
+    /*------------------------ handle back channel logout request start --------------------*/
+    if (authConfig.backChannelAuth.enabled) {
+        var backChannelAuthHandler = Packages.org.wso2.carbon.cloud.back.channel.BackChannelAuthHandler;
+        var issuer = authConfig.backChannelAuth.issuer;
+        if (backChannelAuthHandler.canLogout(issuer, samlRequest)) {
+            log.info("Invalidate back channel authenticated session");
```
Comment (danushkaf) [inline jaggeryapps/admin/acs.jag]: Reason for this being in info level? better to add debug data to log and make it debug

**Review by danushkaf — CHANGES_REQUESTED** [review]

File: jaggeryapps/admin/acs.jag (hunk 7 above)
Comment (Dilhasha) [reply jaggeryapps/admin/acs.jag]: I will change it. Initially I didn't change it because it was already existing in the code.

**Review by danushkaf — APPROVED** [review]

**Review by malithie — APPROVED** [review]

**Pull request merged**